README.md
# Nimbus JOSE + JWT

* The most popular and robust Java 7+ and Android library for JSON Web Tokens 
  (JWT)
* Supports all standard signature (JWS) and encryption (JWE) algorithms, 
  including recent developments such secp256k1, ECDH-1PU and XC20P
* Open source Apache 2.0 license

Check out the [library homepage](https://connect2id.com/products/nimbus-jose-jwt) 
for more info and examples.


## Full compact JOSE and JWT support

Create, parse and process:

* JOSE secured objects:
  * With compact URL-safe serialisation:
    * JSON Web Signature (JWS) objects
    * JSON Web Encryption (JWE) objects
    * Unsecured (alg=none) JOSE objects
    * Signed, encrypted and unsecured JSON Web Tokens (JWTs)
* With JSON serialisation:
  * JWS JSON objects with one or more signatures
  * JWE JSON objects with one or more recipients
* JSON Web Key (JWK) objects and JWK sets


## Supported JOSE algorithms

The library handles the following JOSE algorithms:

* HMAC integrity protection: HS256, HS384 and HS512
* RSASSA-PKCS1-V1_5 signatures: RS256, RS384 and RS512
* RSASSA-PSS signatures: PS256, PS384 and PS512
* EC signatures: ES256, ES256K, ES384, ES512 and EdDSA
* Key encryption with RSAES-PKCS1-V1_5: RSA1_5 (deprecated)
* Key encryption with RSAES OAEP: RSA-OAEP, RSA-OAEP-256 and RSA-OAEP-512
* Key encryption with AES key wrap: A128KW, A192KW and A256KW
* Key encryption with AES GCM: A128CGMKW, A192CGMKW and A256CGMKW
* Direct shared symmetric key encryption: dir
* Key agreement with Elliptic Curve Diffie-Hellman Ephemeral Static: ECDH-ES,
  ECDH-ES+A128KW, ECDH-ES+A192KW and ECDH-ES+A256KW
* Public key authenticated encryption utilising the One-Pass Unified Model for 
  Elliptic Curve Diffie-Hellman key agreement: ECDH-1PU, ECDH-1PU+A128KW, 
  ECDH-1PU+A128KW, ECDH-1PU+A256KW
* Password-based encryption: PBES2-HS256+A128KW, PBES2-HS384+A192KW and
  PBES2-HS512+A256KW
* Content encryption with AES_CBC_HMAC_SHA2: A128CBC-HS256, A192CBC-HS384,
  A256CBC-HS512, the deprecated A128CBC+HS256 and A256CBC+HS512 are also
  supported
* Content encryption with AES GCM: A128GCM, A192GCM and A256GCM
* Content encryption with extended nonce ChaCha20-Poly1305: XC20P
* JWE Compression with DEFLATE.


## Supported IETF standards and drafts

* RFC 7515 - JSON Web Signature (JWS)
* RFC 7516 - JSON Web Encryption (JWE)
* RFC 7517 - JSON Web Key (JWK)
* RFC 7518 - JSON Web Algorithms (JWA)
* RFC 7519 - JSON Web Token (JWT)
* RFC 7520 - Examples of Protecting Content Using JSON Object Signing and
  Encryption (JOSE)
* RFC 7165 - Use Cases and Requirements for JSON Object Signing and Encryption
  (JOSE)
* RFC 7638 - JSON Web Key (JWK) Thumbprint
* RFC 7797 - JSON Web Signature (JWS) Unencoded Payload Option
* RFC 8037 - CFRG Elliptic Curve Diffie-Hellman (ECDH) and Signatures in JSON 
  Object Signing and Encryption (JOSE)
* RFC 8812 - CBOR Object Signing and Encryption (COSE) and JSON Object Signing  
  and Encryption (JOSE) Registrations for Web Authentication (WebAuthn) 
  Algorithms
* draft-ietf-oauth-jwk-thumbprint-uri-01 - JSON Web Key (JWK) Thumbprint URI
* draft-madden-jose-ecdh-1pu-04 - Public Key Authenticated Encryption for JOSE: 
  ECDH-1PU
* draft-amringer-jose-chacha-02 - Chacha derived AEAD algorithms in JSON Object 
  Signing and Encryption (JOSE) (note, support for XC20P only)
* draft-irtf-cfrg-xchacha-03 - XChaCha: eXtended-nonce ChaCha and 
  AEAD_XChaCha20_Poly1305


## System requirements and dependencies

The Nimbus JOSE+JWT library requires Java 7+ and has minimal dependencies.

* (optional) BouncyCastle as an alternative JCA provider. Must not be imported
  together with the BouncyCastle FIPS provider!
* (optional) BouncyCastle FIPS as a FIPS 140-2, Level 1 compliant JCA provider.
  Must not be imported together with the plain BouncyCastle provider!
* (optional) Tink for EdDSA with Ed25519, ECDH with X25519 and content 
  encryption with XC20P.


For Maven add:

```
<dependency>
    <groupId>com.nimbusds</groupId>
    <artifactId>nimbus-jose-jwt</artifactId>
    <version>[ version ]</version>
</dependency>
```

where `[ version ]` is the latest stable version.


## Issues / suggestions

To post bug reports and suggestions:

<https://bitbucket.org/connect2id/nimbus-jose-jwt/issues>


## SonarCloud Status

[![Quality Gate Status](https://sonarcloud.io/api/project_badges/measure?project=connect2id_nimbus-jose-jwt&metric=alert_status)](https://sonarcloud.io/dashboard?id=connect2id_nimbus-jose-jwt)
[![Security Rating](https://sonarcloud.io/api/project_badges/measure?project=connect2id_nimbus-jose-jwt&metric=security_rating)](https://sonarcloud.io/dashboard?id=connect2id_nimbus-jose-jwt)
[![Vulnerabilities](https://sonarcloud.io/api/project_badges/measure?project=connect2id_nimbus-jose-jwt&metric=vulnerabilities)](https://sonarcloud.io/dashboard?id=connect2id_nimbus-jose-jwt)
[![Bugs](https://sonarcloud.io/api/project_badges/measure?project=connect2id_nimbus-jose-jwt&metric=bugs)](https://sonarcloud.io/dashboard?id=connect2id_nimbus-jose-jwt)
[![Coverage](https://sonarcloud.io/api/project_badges/measure?project=connect2id_nimbus-jose-jwt&metric=coverage)](https://sonarcloud.io/dashboard?id=connect2id_nimbus-jose-jwt)
[![Lines of Code](https://sonarcloud.io/api/project_badges/measure?project=connect2id_nimbus-jose-jwt&metric=ncloc)](https://sonarcloud.io/dashboard?id=connect2id_nimbus-jose-jwt)

## Twitter

Follow updates and new releases on Twitter:

<https://twitter.com/connect2id>



CHANGELOG.txt
version 1.0 (2012-03-01)
	* First version based on the OpenInfoCard JWT, JWS and JWE code base.

version 1.1 (2012-03-06)
	* Introduces type-safe enumeration of the JSON Web Algorithms (JWA).
	* Refactors the JWT class.

version 1.2 (2012-03-08)
	* Moves JWS and JWE code into separate classes.

version 1.3 (2012-03-09)
	* Switches to Apache Commons Codec for Base64URL encoding and decoding
	* Consolidates the crypto utilities within the package.
	* Introduces a JWT content serialiser class.

version 1.4 (2012-03-09)
	* Refactoring of JWT class and JUnit tests.

version 1.5 (2012-03-18)
	* Switches to JSON Smart for JSON serialisation and parsing.
	* Introduces claims set class with JSON objects, string, Base64URL and
	  byte array views.

version 1.6 (2012-03-20)
	* Creates class for representing, serialising and parsing JSON Web Keys
	  (JWK).
	* Introduces separate class for representing JWT headers.

version 1.7 (2012-04-01)
	* Introduces separate classes for plain, JWS and JWE headers.
	* Introduces separate classes for plain, signed and encrypted JWTs.
	* Removes the JWTContent class.
	* Removes password-based (PE820) encryption support.

version 1.8 (2012-04-03)
	* Adds support for the ZIP JWE header parameter.
	* Removes unsupported algorithms from the JWA enumeration.

version 1.9 (2012-04-03)
	* Renames JWEHeader.{get|set}EncryptionAlgorithm() to
	  JWEHeader.{get|set}EncryptionMethod().

version 1.9.1 (2012-04-03)
	* Upgrades JSON Smart JAR to 1.1.1.

version 1.10 (2012-04-14)
	* Introduces serialize() method to base abstract JWT class.

version 1.11 (2012-05-13)
	* JWT.serialize() throws checked JWTException instead of
	  IllegalStateException.

version 1.12 (2012-05-21)
	* Updates the alternative JWT header type ("typ") parameter value to
	  "urn:ietf:params:oauth:token-type:jwt" (to reflect changes in
	  draft-jones-json-web-token-10).
	* Renames the JWS and JWE header parameter "jpk" to "jwk" (to reflect
	  changes in draft-ietf-jose-json-web-signature-02 and
	  draft-ietf-jose-json-web-encryption-02).
	* Changes the JWE compression algorithm to DEFLATE (to reflect changes
	  in draft-ietf-jose-json-web-encryption-02).
	* Renames JWK class to JWKSet and JWKKeyObject class to JWK (to reflect
	  changes in draft-ietf-jose-json-web-key-02).
	* Adds toJSONObject() method to the ReadOnlyHeader interface.
	* Updates JavaDocs.

version 2.0 (2012-09-29)
	* Full rewrite of 1.* code to decouple plain/JWS/JWE/JWT object
	  representation from JWA implementations.
	* JWS code updated to draft-ietf-jose-json-web-signature-05.
	* JWE code updated to draft-ietf-jose-json-web-encryption-05.
	* JWK code updated to draft-ietf-jose-json-web-key-05.
	* JWA code updated to draft-ietf-jose-json-web-algorithms-05.
	* JWT code updated to draft-ietf-oauth-json-web-token-03.
	* Removes existing JWE algorithm implementations since no longer
	  compatible with newer -05 specs (to be reimplemented).

version 2.0.1 (2012-09-29)
	* Adds JWS object use example.

version 2.1 (2012-10-07)
	* Changes naming from "verifier" to "validator" to conform with JWS
	  terminology.
	* Refactors library to enforce header parameter checking of signed JWS
	  and encrypted JWE objects.
	* Simplifies package tree.

version 2.1.1 (2012-10-09)
	* Upgrades BouncyCastle.org JAR to 1.47.
	* Removes stale RSASSAVerifier.

version 2.2 (2012-10-16)
	* Updates the library to the latest JWA/JWS/JWE/JWK/JWT -06 draft suite.

version 2.3 (2012-10-23)
	* Naming changes (validation, verification) according to RFC 4949.
	* Makes PlainObject, JWSObject and JWEObject thread-safe.
	* Adds @Immutable and @ThreadSafe annotations (JCIP Annotations JAR).
	* Adds JOSEObject.getParsedParts() and getParsedString() methods to
	  get the original parsed Base64URL parts / string representation.

version 2.4 (2012-11-26)
	* Adds Maven POM support.

version 2.5 (2012-11-26)
	* Updates the library to the latest JWA/JWS/JWE/JWK -07 draft suite.

version 2.6 (2012-11-19)
	* Adds support for additional custom JSON Web Key (JWK) set members.

version 2.7 (2012-12-09)
	* Adds toBaseURL() method to ReadOnlyHeader interface.
	* Fixes assignment bug in JWEObject Base64URL parts constructor.

version 2.8 (2012-12-17)
	* Adds getParsedParts() and getParsedString() methods to JWT interface.

version 2.9 (2013-01-08)
	* Updates the library to the latest JWA/JWS/JWE/JWK -08 draft suite.
	* Updates the library to the latest JWT -06 draft.
	* Minor code refactoring and clean up.

version 2.10 (2013-01-15)
	* Fixes incorrect null argument treatment in overridden Object.equals()
	  methods.
	* Renames ReadOnlyClaimsSet to ReadOnlyJWTClaimsSet and ClaimsSet to
	  JWTClaimsSet.

version 2.10.1 (2013-01-16)
	* Adds Maven support.
	* Fixes bug in build.xml script.

version 2.11.0 (2013-02-22)
	* Adds RSA1_5 / RSA OAEP with AES-GCM encryption and decryption.
	* Refactors JWTClaimsSet.

version 2.12.0 (2013-03-21)
	* Adds JSON Private and Symmetric Key (JPSK) support.
	* Extends HeaderFilter to allow setting of accepted header parameters.
	* Fixes ECDSA signature R + S part processing for ES384 and ES512.
	* Removes Ant build script.

version 2.13.0 (2013-03-28)
	* Adds support for A128CBC+HS256 and A256CBC+HS512 encryption methods.
	* Implements ECKey export and import.
	* Adds JWE DEFLATE support.
	* Implements basic MMA attack protection for RSA1_5 JWE algorithm (issue
	  #23).
	* Refactors JOSE header filters.
	* Refactors crypto package.
	* Adds support to translate JWKs to and from native Java key objects.

version 2.13.1 (2013-04-15)
	* Fixes incorrect time unit representation of "iat", "exp" and "nbf" JWT
	  claims.
	* If the JOSE header was parsed Header.toBase64URL() returns the
	  original Base64URL.

version 2.14 (2013-04-16)
	* Adds support for direct JWE encryption and decryption.
	* Adds JWKSet method to lookup a key by its key ID (kid).
	* Fixes DEFLATE compression by adding nowrap flag to omit headers and
	  CRC fields.

version 2.15 (2013-05-07)

src/main/java/com/nimbusds/jose/jca/JCASupport.java
/*
 * nimbus-jose-jwt
 *
 * Copyright 2012-2016, Connect2id Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
 * this file except in compliance with the License. You may obtain a copy of the
 * License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed
 * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */

package com.nimbusds.jose.jca;


import java.security.NoSuchAlgorithmException;
import java.security.Provider;
import java.security.Security;
import javax.crypto.Cipher;
import javax.crypto.Mac;
import javax.crypto.NoSuchPaddingException;

import com.nimbusds.jose.*;
import com.nimbusds.jose.crypto.impl.ECDSA;
import com.nimbusds.jose.crypto.impl.RSASSA;


/**
 * Java Cryptography Architecture (JCA) support helper.
 *
 * @version 2022-05-16
 */
public final class JCASupport {


	/**
	 * Checks if unlimited cryptographic strength is supported. If not
	 * download the appropriate jurisdiction policy files for your Java
	 * edition:
	 *
	 * <p><a href="http://www.oracle.com/technetwork/java/javase/downloads/jce-7-download-432124.html">JCE Unlimited Strength Jurisdiction Policy Files for Java 7</a>
	 *
	 * <p><a href="http://www.oracle.com/technetwork/java/javase/downloads/jce8-download-2133166.html">JCE Unlimited Strength Jurisdiction Policy Files for Java 8</a>
	 *
	 * @return {@code true} if unlimited cryptographic strength is
	 *         supported, {@code false} if not.
	 */
	public static boolean isUnlimitedStrength() {

		try {
			return Cipher.getMaxAllowedKeyLength("AES") >= 256;
		} catch (NoSuchAlgorithmException e) {
			return false;
		}
	}
	
	
	/**
	 * Checks if the specified JOSE algorithm is supported by the default
	 * system JCA provider(s).
	 *
	 * @param alg The JOSE algorithm. Must not be {@code null}.
	 *
	 * @return {@code true} if the JOSE algorithm is supported, else
	 *         {@code false}.
	 */
	public static boolean isSupported(final Algorithm alg) {
		
		if (alg instanceof JWSAlgorithm) {
			return isSupported((JWSAlgorithm)alg);
		}
		if (alg instanceof JWEAlgorithm) {
			return isSupported((JWEAlgorithm)alg);
		}
		if (alg instanceof EncryptionMethod) {
			return isSupported((EncryptionMethod)alg);
		}
		throw new IllegalArgumentException("Unexpected algorithm class: " + alg.getClass().getCanonicalName());
	}
	
	
	/**
	 * Checks if a JOSE algorithm is supported by the specified JCA
	 * provider.
	 *
	 * @param alg      The JOSE algorithm. Must not be {@code null}.
	 * @param provider The JCA provider. Must not be {@code null}.
	 *
	 * @return {@code true} if the JOSE algorithm is supported, else
	 *         {@code false}.
	 */
	public static boolean isSupported(final Algorithm alg, final Provider provider) {
		
		if (alg instanceof JWSAlgorithm) {
			return isSupported((JWSAlgorithm)alg, provider);
		}
		if (alg instanceof JWEAlgorithm) {
			return isSupported((JWEAlgorithm)alg, provider);
		}
		if (alg instanceof EncryptionMethod) {
			return isSupported((EncryptionMethod)alg, provider);
		}
		throw new IllegalArgumentException("Unexpected algorithm class: " + alg.getClass().getCanonicalName());
	}


	/**
	 * Checks if the specified JWS algorithm is supported by the default
	 * system JCA provider(s).
	 *
	 * @param alg The JWS algorithm. Must not be {@code null}.
	 *
	 * @return {@code true} if the JWS algorithm is supported, else
	 *         {@code false}.
	 */
	public static boolean isSupported(final JWSAlgorithm alg) {
		
		if (alg.getName().equals(Algorithm.NONE.getName())) {
			return true;
		}

		for (Provider p: Security.getProviders()) {

			if (isSupported(alg, p)) {
				return true;
			}
		}

		return false;
	}


	/**
	 * Checks if a JWS algorithm is supported by the specified JCA
	 * provider.
	 *
	 * @param alg      The JWS algorithm. Must not be {@code null}.
	 * @param provider The JCA provider. Must not be {@code null}.
	 *
	 * @return {@code true} if the JWS algorithm is supported, else
	 *         {@code false}.
	 */
	public static boolean isSupported(final JWSAlgorithm alg, final Provider provider) {

		if (JWSAlgorithm.Family.HMAC_SHA.contains(alg)) {
			String jcaName;
			if (alg.equals(JWSAlgorithm.HS256)) {
				jcaName = "HMACSHA256";
			} else if (alg.equals(JWSAlgorithm.HS384)) {
				jcaName = "HMACSHA384";
			} else if (alg.equals(JWSAlgorithm.HS512)) {
				jcaName = "HMACSHA512";
			} else {
				return false;
			}
			try {
				Mac.getInstance(jcaName, provider);
			} catch (NoSuchAlgorithmException e) {
				return false;
			}
			return true;
		}

		if (JWSAlgorithm.Family.RSA.contains(alg)) {
			try {
				RSASSA.getSignerAndVerifier(alg, provider);
			} catch (JOSEException e) {
				return false;
			}
			return true;
		}

		if (JWSAlgorithm.Family.EC.contains(alg)) {
			try {
				ECDSA.getSignerAndVerifier(alg, provider);
			} catch (JOSEException e) {
				return false;
			}
			return true;
		}

		return false;
	}


	/**
	 * Checks if the specified JWE algorithm is supported by the default
	 * system JCA provider(s).
	 *
	 * @param alg The JWE algorithm. Must not be {@code null}.
	 *
	 * @return {@code true} if the JWE algorithm is supported, else
	 *         {@code false}.
	 */
	public static boolean isSupported(final JWEAlgorithm alg) {

		for (Provider p: Security.getProviders()) {

			if (isSupported(alg, p)) {
				return true;
			}
		}

		return false;
	}


	/**
	 * Checks if a JWE algorithm is supported by the specified JCA
	 * provider.
	 *
	 * @param alg      The JWE algorithm. Must not be {@code null}.
	 * @param provider The JCA provider. Must not be {@code null}.
	 *
	 * @return {@code true} if the JWE algorithm is supported, else
	 *         {@code false}.
	 */
	public static boolean isSupported(final JWEAlgorithm alg, final Provider provider) {

		String jcaName;

		if (JWEAlgorithm.Family.RSA.contains(alg)) {
			if (alg.equals(JWEAlgorithm.RSA1_5)) {
				jcaName = "RSA/ECB/PKCS1Padding";
			} else if (alg.equals(JWEAlgorithm.RSA_OAEP)) {
				jcaName = "RSA/ECB/OAEPWithSHA-1AndMGF1Padding";
			} else if (alg.equals(JWEAlgorithm.RSA_OAEP_256)) {
				jcaName = "RSA/ECB/OAEPWithSHA-256AndMGF1Padding";
			} else if (alg.equals(JWEAlgorithm.RSA_OAEP_512)) {
				jcaName = "RSA/ECB/OAEPWithSHA-512AndMGF1Padding";
			} else {
				return false;
			}

			// Do direct test
			try {
				Cipher.getInstance(jcaName, provider);
			} catch (NoSuchAlgorithmException | NoSuchPaddingException e) {
				return false;
			}
			return true;
		}

		if (JWEAlgorithm.Family.AES_KW.contains(alg)) {
			return provider.getService("Cipher", "AESWrap") != null;
		}

		if (JWEAlgorithm.Family.ECDH_ES.contains(alg)) {
			return provider.getService("KeyAgreement", "ECDH") != null;
		}

		if (JWEAlgorithm.Family.AES_GCM_KW.contains(alg)) {
			// Do direct test
			try {
				Cipher.getInstance("AES/GCM/NoPadding", provider);
			} catch (NoSuchAlgorithmException | NoSuchPaddingException e) {
				return false;
			}
			return true;
		}

		if (JWEAlgorithm.Family.PBES2.contains(alg)) {
			String hmac;
			if (alg.equals(JWEAlgorithm.PBES2_HS256_A128KW)) {
				hmac = "HmacSHA256";
			} else if (alg.equals(JWEAlgorithm.PBES2_HS384_A192KW)) {
				hmac = "HmacSHA384";
			} else {
				hmac = "HmacSHA512";
			}
			return provider.getService("KeyGenerator", hmac) != null;
		}
		
		return JWEAlgorithm.DIR.equals(alg); // Always supported
	}


	/**
	 * Checks if the specified JWE encryption method is supported by the
	 * default system JCA provider(s).
	 *
	 * @param enc The JWE encryption method. Must not be {@code null}.
	 *
	 * @return {@code true} if the JWE algorithm is supported, else
	 *         {@code false}.
	 */
	public static boolean isSupported(final EncryptionMethod enc) {

		for (Provider p: Security.getProviders()) {

			if (isSupported(enc, p)) {
				return true;
			}
		}

		return false;
	}


	/**
	 * Checks if a JWE encryption method is supported by the specified
	 * JCA provider.
	 *
	 * @param enc      The JWE encryption method. Must not be {@code null}.
	 * @param provider The JCA provider. Must not be {@code null}.
	 *
	 * @return {@code true} if the JWE encryption method is supported, else
	 *         {@code false}.
	 */
	public static boolean isSupported(final EncryptionMethod enc, final Provider provider) {

		if (EncryptionMethod.Family.AES_CBC_HMAC_SHA.contains(enc)) {
			// Do direct test
			try {
				Cipher.getInstance("AES/CBC/PKCS5Padding", provider);
			} catch (NoSuchAlgorithmException | NoSuchPaddingException e) {
				return false;
			}
			// Check hmac
			String hmac;
			if (enc.equals(EncryptionMethod.A128CBC_HS256)) {
				hmac = "HmacSHA256";
			} else if (enc.equals(EncryptionMethod.A192CBC_HS384)) {
				hmac = "HmacSHA384";
			} else {
				hmac = "HmacSHA512";
			}
			return provider.getService("KeyGenerator", hmac) != null;
		}

		if (EncryptionMethod.Family.AES_GCM.contains(enc)) {
			// Do direct test
			try {
				Cipher.getInstance("AES/GCM/NoPadding", provider);
			} catch (NoSuchAlgorithmException | NoSuchPaddingException e) {
				return false;
			}
			return true;
		}

		return false;
	}


	/**
	 * Prevents public instantiation.
	 */
	private JCASupport() {

	}
}
